superset/views/chart/views.py

```diff
@@ -40,22 +41,22 @@ class SliceModelView(
         RouteMethod.API_DELETE,
     }
 
-    def pre_add(self, item):
+    def pre_add(self, item: "SliceModelView") -> None:
         utils.validate_json(item.params)
 
-    def pre_update(self, item):
+    def pre_update(self, item: "SliceModelView") -> None:
         utils.validate_json(item.params)
         check_ownership(item)
 
-    def pre_delete(self, item):
+    def pre_delete(self, item: "SliceModelView") -> None:
         check_ownership(item)
 
     @expose("/add", methods=["GET", "POST"])
     @has_access
-    def add(self):
-        datasources = ConnectorRegistry.get_all_datasources(db.session)
+    def add(self) -> FlaskResponse:
         datasources = [
-            {"value": str(d.id) + "__" + d.type, "label": repr(d)} for d in datasources
+            {"value": str(d.id) + "__" + str(d.type), "label": repr(d)}
```

Comment:
Note I need to wrap `d.type` in a `str(...)` as this can be optionally `None` (for right or wrong) per [here](https://github.com/apache/incubator-superset/blob/e5f5eed42504185723b3caf5b2042db333de32c7/superset/connectors/base/models.py#L41).

Reply:
I think we can treat `d.type` as being an abstract property that is always set. Couldn't we just change that to be `type: str`, or optionally make it
```
@property
def type(self) -> str:
    raise NotImplementedError()
```

Reply:
Thanks for the suggestion @villebro. I think in the long term these abstract base classes should use [ABC](https://docs.python.org/3/library/abc.html) or similar and definitely not set defaults like they have.